R/derive_extreme_event.R

```diff
@@ -0,0 +1,205 @@
+#'
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#'
+#' adqs <- tribble(
+#'   ~USUBJID, ~PARAMCD,       ~AVALC,        ~ADY,
+#'   "1",      "NO SLEEP",     "N",              1,
+#'   "1",      "WAKE UP",      "N",              2,
+#'   "1",      "FALL ASLEEP",  "N",              3,
+#'   "2",      "NO SLEEP",     "N",              1,
+#'   "2",      "WAKE UP",      "Y",              2,
+#'   "2",      "WAKE UP",      "Y",              3,
+#'   "2",      "FALL ASLEEP",  "N",              4,
+#'   "3",      "NO SLEEP",     NA_character_,    1
+#' )
+#'
+#' # Add a new record for each USUBJID storing the the worst sleeping problem.
+#' derive_extreme_event(
+#'   adqs,
+#'   by_vars = exprs(USUBJID),
+#'   events = list(
+#'     event(
+#'       condition = PARAMCD == "NO SLEEP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "No sleep", AVAL = 1)
+#'     ),
+#'     event(
+#'       condition = PARAMCD == "WAKE UP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "Waking up more than three times", AVAL = 2)
+#'     ),
+#'     event(
+#'       condition = PARAMCD == "FALL ASLEEP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "More than 30 mins to fall asleep", AVAL = 3)
+#'     ),
+#'     event(
+#'       condition = all(AVALC == "N"),
+#'       set_values_to = exprs(
+#'         AVALC = "No sleeping problems", AVAL = 4
+#'       )
+#'     ),
+#'     event(
+#'       condition = TRUE,
+#'       set_values_to = exprs(AVALC = "Missing", AVAL = 99)
+#'     )
+#'   ),
+#'   order = exprs(ADY),
+#'   mode = "last",
+#'   set_values_to = exprs(
+#'     PARAMCD = "WSP",
+#'     PARAM = "Worst Sleeping Problems"
+#'   )
+#' )
+derive_extreme_event <- function(dataset,
+                                 by_vars = NULL,
+                                 events,
+                                 order,
+                                 mode,
+                                 check_type = "warning",
+                                 set_values_to) {
+  # Check input parameters
+  assert_vars(by_vars, optional = TRUE)
+  assert_list_of(events, "event")
+  assert_order_vars(order)
+  assert_data_frame(
+    dataset,
+    required_vars = by_vars
+  )
+  mode <- assert_character_scalar(mode, values = c("first", "last"), case_sensitive = FALSE)
+  check_type <-
+    assert_character_scalar(
+      check_type,
+      values = c("none", "warning", "error"),
+      case_sensitive = FALSE
+    )
+  assert_varval_list(set_values_to)
+
+  # Create new observations
+  ## Create a dataset (selected_records) from `events`
+  condition_ls <- map(events, "condition")
+  set_values_to_ls <- map(events, "set_values_to")
+  event_order <- map(seq_len(length(events)), function(x) x)
+
+  selected_records_ls <- purrr::pmap(
+    list(condition_ls, set_values_to_ls, event_order),
+    function(x, y, z) {
+      dataset %>%
+        group_by(!!!by_vars) %>%
+        filter(!!enexpr(x)) %>%
+        mutate(!!!y, event_no = z)
```

Comment:
Could you use `get_new_tmp_var()` to create a name for the temporary variable? Then it can not happen that an existing variable is accidentally overwritten.